selfdrive/ui/soundd.py

```diff
@@ -0,0 +1,152 @@
+}
+
+def check_controls_state(sm):
+  controls_missing = time.monotonic() - sm.rcv_time['controlsState']
+
+  if controls_missing > CONTROLS_TIMEOUT:
+    if sm['controlsState'].enabled and (controls_missing - CONTROLS_TIMEOUT) < 10:
+      return False
+
+  return True
+
+
+class Soundd:
+  def __init__(self):
+    self.load_sounds()
+
+    self.current_alert = AudibleAlert.none
+    self.current_volume = MIN_VOLUME
+    self.current_sound_frame = 0
+
+  def load_sounds(self):
+    self.loaded_sounds: Dict[int, np.ndarray] = {}
+
+    # Load all sounds
+    for sound in sound_list:
+      filename, play_count, volume = sound_list[sound]
+
+      wavefile = wave.open(BASEDIR + "/selfdrive/assets/sounds/" + filename, 'r')
+
+      assert wavefile.getnchannels() == 1
+      assert wavefile.getsampwidth() == 2
+      assert wavefile.getframerate() == SAMPLE_RATE
+
+      length = wavefile.getnframes()
+      self.loaded_sounds[sound] = np.frombuffer(wavefile.readframes(length), dtype=np.int16).astype(np.float32) / 32767
+
+  def get_sound_data(self, frames): # get "frames" worth of data from the current alert sound, looping when required
+
+    ret = np.zeros(frames, dtype=np.float32)
+
+    if self.current_alert != AudibleAlert.none:
+      num_loops = sound_list[self.current_alert][1]
+      sound_data = self.loaded_sounds[self.current_alert]
+      written_frames = 0
+
+      current_sound_frame = self.current_sound_frame % len(sound_data)
+      loops = self.current_sound_frame // len(sound_data)
+
+      while written_frames < frames and (num_loops is None or loops < num_loops):
+        available_frames = sound_data.shape[0] - current_sound_frame
+        frames_to_write = min(available_frames, frames - written_frames)
+        ret[written_frames:written_frames+frames_to_write] = sound_data[current_sound_frame:current_sound_frame+frames_to_write]
+        written_frames += frames_to_write
+        self.current_sound_frame += frames_to_write
+
+    return ret * self.current_volume
+
+  def callback(self, data_out: np.ndarray, frames: int, time, status) -> None:
+    if status:
+      cloudlog.warning(f"soundd stream over/underflow: {status}")
+    data_out[:frames, 0] = self.get_sound_data(frames)
+
+  def update_alert(self, new_alert):
+    current_alert_played_once = self.current_alert == AudibleAlert.none or self.current_sound_frame > len(self.loaded_sounds[self.current_alert])
+    if self.current_alert != new_alert and (new_alert != AudibleAlert.none or current_alert_played_once):
+      self.current_alert = new_alert
+      self.current_sound_frame = 0
+
+  def get_audible_alert(self, sm):
+    if sm.updated['controlsState']:
+      new_alert = sm['controlsState'].alertSound.raw
+      self.update_alert(new_alert)
+    elif not check_controls_state(sm):
+      self.update_alert(AudibleAlert.warningImmediate)
+    else:
+      self.update_alert(AudibleAlert.none)
```

Comment:
I think you want to remove this branch. looks like if soundd was running at >100Hz for example, it would cut the sounds immediately after starting them

Reply:
this is for stopping the timeout alert, I made a flag to make that more clear and also more robust if we are running >100hz